### PR TITLE
Client: Implement turns

### LINKED ISSUE #195: Client: Implement player turns
After #194 implement turns in client

Additionally:
- Add visual indicator showing whose turn it is, maybe with nicknames as well
- when opponent makes a move, the corresponding piece should be animated, can be simple for now (lift, jump to new position, drop)
- player shouldn't be able to interact with opponents pieces (they shouldn't light up on hover

### PR DESCRIPTION
**Pull request by Fruktus**
Closes #195

### CODE REVIEW

File: QuadradiusR/scripts/network_handler.gd [hunk 1]
```diff
@@ -40,16 +46,30 @@ func _create_user_1(message: Message, args: Dictionary):
...
+
+	if is_authorized:
+		token = message.body['token'] if is_authorized else ""
+		rest_api.get_user_me(self.token, funcref(self, "_authorize_user_2"), args)
+	else:
+		args['cb'].call_func(is_authorized, args['user_data'], message)
+
+func _authorize_user_2(message: Message, args: Dictionary):
+	username = message.body['username']
+	user_id = message.body['id']
```
Comment (Observer46) [inline QuadradiusR/scripts/network_handler.gd:66]: `self.username`, `self.user_id` or maybe put in a new `Context` class

File: QuadradiusR/scripts/network_handler.gd [hunk 2]
```diff
@@ -63,9 +83,41 @@ func _join_lobby_1(message: Message, args: Dictionary):
...
+# INVITE PLAYER #
+# # # # # # # # #
 func invite_player(opponent_uuid: String):
 	rest_api.invite_player(token, opponent_uuid)
+
+
+# # # # # # # # # # # # #
+# ACCEPT AND JOIN GAME  #
+# # # # # # # # # # # # #
+func accept_and_join_game(game_invite_id, cb: FuncRef):
```
Comment (Observer46) [inline QuadradiusR/scripts/network_handler.gd:99]: Type hinting if you'd like

File: QuadradiusR/scripts/network_handler.gd [hunk 3]
```diff
@@ -63,9 +83,41 @@ func _join_lobby_1(message: Message, args: Dictionary):
...
 
-	
 	args['cb'].call_func()
 
 
+# # # # # # # # #
+# INVITE PLAYER #
+# # # # # # # # #
 func invite_player(opponent_uuid: String):
 	rest_api.invite_player(token, opponent_uuid)
```
Comment (Observer46) [inline QuadradiusR/scripts/network_handler.gd:93]: `opponent_uuid` &rarr; `opponent_id`

File: QuadradiusR/scripts/network_handler.gd [hunk 4]
```diff
@@ -63,9 +83,41 @@ func _join_lobby_1(message: Message, args: Dictionary):
...
+
+
+# # # # # # # # # # # # #
+# ACCEPT AND JOIN GAME  #
+# # # # # # # # # # # # #
+func accept_and_join_game(game_invite_id, cb: FuncRef):
+	rest_api.accept_game_invite(self.token, game_invite_id, funcref(self, "_accept_and_join_game_1"), {"game_invite_id": game_invite_id, 'cb': cb})
+
+func _accept_and_join_game_1(message: Message, args: Dictionary):
+	var game_id = message.headers[0].substr(16, len(message.headers[0]))
```
Comment (Observer46) [inline QuadradiusR/scripts/network_handler.gd:103]: Need to fix that thing with headers due their order

File: QuadradiusR/scripts/network_handler.gd [hunk 5]
```diff
@@ -63,9 +83,41 @@ func _join_lobby_1(message: Message, args: Dictionary):
...
+	join_game(game_id, args['cb'])
+
+
+# # # # # # #
+# JOIN GAME #
+# # # # # # #
+func join_game(game_id, cb: FuncRef):
```
Comment (Observer46) [inline QuadradiusR/scripts/network_handler.gd:110]: Type hinting if you would, for that is our curse

File: QuadradiusR/scripts/network_handler.gd [hunk 6]
```diff
@@ -40,16 +46,30 @@ func _create_user_1(message: Message, args: Dictionary):
...
+	if is_authorized:
+		token = message.body['token'] if is_authorized else ""
+		rest_api.get_user_me(self.token, funcref(self, "_authorize_user_2"), args)
+	else:
+		args['cb'].call_func(is_authorized, args['user_data'], message)
+
+func _authorize_user_2(message: Message, args: Dictionary):
+	username = message.body['username']
+	user_id = message.body['id']
+	args['cb'].call_func(true, args['user_data'], message)
```
Comment (Observer46) [inline QuadradiusR/scripts/network_handler.gd:67]: We need an interface/documentation about callback if we use parameteres.eed an interface/documentation about callback if we use parameteres.

File: QuadradiusR/scripts/player_list.gd [hunk 7]
```diff
@@ -66,6 +66,6 @@ func _lobby_left(lobby_id: String, user_id: String):
 	uuid_to_active_invites.erase(user_id)
 	remove_player(user_id)
 
-func _game_invite_received(game_id, from_id, from_username, subject_id, subject_username):
-	uuid_to_active_invites[from_id] = game_id
+func _game_invite_received(game_invite_id, from_id, from_username, subject_id, subject_username):
```
Comment (Observer46) [inline QuadradiusR/scripts/player_list.gd:69]: Typie hinting

File: QuadradiusR/scripts/ws_api.gd [hunk 8]
```diff
@@ -125,11 +142,19 @@ func connect_to(url: String, token):
...
+func close(reason: String = ""):
+	ws.disconnect_from_host(1000, reason)
+
+
 func subscribe_to(topic: String):
 	var query = {"op": SUBSCRIBE, "d": {"topic": topic}}
 	_send_data(query)
 
 
+func make_move(piece_id, tile_id):
```
Comment (Observer46) [inline QuadradiusR/scripts/ws_api.gd:154]: type hinting if you can 😸

**Review by Observer46 — CHANGES_REQUESTED** [review]
some strokes were made but overall it's gucci!

File: QuadradiusR/scripts/board.gd [hunk 9]
```diff
@@ -33,21 +32,28 @@ func _init_tiles():
...
-		
-		_get_child_at_idx(i).set_slot(player1_torus)
-		_get_child_at_idx(board_size.x * board_size.y - 1 - i).set_slot(player2_torus)
+func _init_toruses(player_pieces: Dictionary):
+	for piece_id in player_pieces.keys():
+		var piece_data = player_pieces[piece_id]
+		var tile = NetworkHandler.game_state.get_tile_by_id(piece_data['tile_id'], self)
+		var color = Torus.COLORS.RED if piece_data['owner_id'] == NetworkHandler.user_id else Torus.COLORS.BLUE
+		var torus = torus_template.instance().init(self, tile, piece_data['owner_id'], piece_id, color)
+		_get_child_at_pos(tile.tile_pos.x, tile.tile_pos.y).set_slot(torus)
```
[the remark this answers is not in the document]
Comment (Fruktus) [reply QuadradiusR/scripts/board.gd:45]: moved to issue

File: QuadradiusR/scripts/network_handler.gd (hunk 1 above)
Comment (Fruktus) [reply QuadradiusR/scripts/network_handler.gd]: Moved to new class

File: QuadradiusR/scripts/network_handler.gd (hunk 6 above)
Comment (Fruktus) [reply QuadradiusR/scripts/network_handler.gd:64]: Moved to new issue

**Review by kjarosh — APPROVED** [review]

**Pull request merged**